[MRG] Fixing label clamping for LabelPropagation

boechat107 opened the pull request:
### Referenced issues

This PR comes to fix #3550 and #5774. It's similar to PR #3758, which seems to be abandoned.

**Important note**: my development files `hack-dev/*` would be removed before merging this PR.
### Explanations

From the comments in the PR #3758, this PR improves the documentation of the parameter `alpha`, deprecates it for `LabelPropagation` and fix the label clamping (the credit is all to @musically-ut).

From [1] and [3], we can see that `LabelPropagation` doesn't have a clamping factor. In the case of `LabelSpreading`, from [2] and [3] we have 

![image](https://cloud.githubusercontent.com/assets/1699905/14866129/5dae176c-0cc2-11e6-985a-c9d4630e8f41.png)

which means that `alpha = 0` keeps the initial label information `Ŷ(0)`. As suggested by @amueller in #3758, this should be the only possible value for `LabelPropagation`.

In the current implementation, [line 239](https://github.com/scikit-learn/scikit-learn/blob/78a674875e30417afe9621346ba022834c138351/sklearn/semi_supervised/label_propagation.py#L239), `alpha = 0` would not propagate label information to the unlabeled instances, but would propagate label information to the labeled instances. I saw this exact behavior with my data, but I couldn't find yet a simple test case to catch it.
### Some additional comments

In #5774, @musically-ut discussed a very important point about the current implementation, the graph construction. There is no mention about directed graphs in the referenced papers, but it's what the current implementation does ([line 137](https://github.com/scikit-learn/scikit-learn/blob/78a674875e30417afe9621346ba022834c138351/sklearn/semi_supervised/label_propagation.py#L137)). Maybe a specific issue should be created for this.
### Referenced papers

[1] Zhu, Xiaojin, and Zoubin Ghahramani. Learning from labeled and unlabeled data with label propagation. Technical Report CMU-CALD-02-107, Carnegie Mellon University, 2002.
[2] Zhou, Dengyong, et al. "Learning with local and global consistency." Advances in neural information processing systems 16.16 (2004): 321-328.
[3] Bengio, Yoshua, Olivier Delalleau, and Nicolas Le Roux. "Label propagation and quadratic criterion." Semi-supervised learning (2006): 193-216.

boechat107 committed:
Files for my dev environment with Docker

boechat107 committed:
Fixing label clamping (alpha=0 for hard clamping)

boechat107 committed:
Deprecating alpha, fixing its value to zero

boechat107 committed:
Correct way to deprecate alpha for LabelPropagation

The previous way was breaking the test
sklearn.tests.test_common.test_all_estimators

boechat107 committed:
Detailed info for LabelSpreading's alpha parameter

Based on the original paper.

boechat107 committed:
Minor changes in the deprecation message

amueller commented:
Sorry for not reviewing this so far. I think it's an important fix. The maintainers are pretty busy :(

boechat107 commented:
No worries, I understand. ;-)

jnothman commented:
This needs some love soon.

amueller commented:
Can you elaborate on the directed graph issue? The construction creates a symmetric graph matrix, right?

musically-ut commented:
@amueller I don't think the default implementation of `kneighbors_graph` creates symmetric graphs (it would be very odd if it did, since k-neighbors may not be symmetric).

I corrected by making the connectivity matrix symmetric explicitly: https://github.com/musically-ut/semi_supervised/blob/master/semi_supervised/label_propagation.py#L136

Does this answer your question?

amueller reviewed (commented):
Can you please add the test suggested in #5774 (at least)?

amueller commented:
@musically-ut yeah, sorry, of course. I was thinking about distance graphs... getting late over here...

amueller commented:
We might want to think about renaming ``alpha`` to something more intuitive to force users to rethink what they're doing....

musically-ut commented:
@amueller, no worries. Great to see progress on this though!

musically-ut commented:
@amueller I'm not inclined to change the name of the parameter, it is called `alpha` in all the papers regarding semi-supervised learning that I have seen.

amueller commented:
@musically-ut we should have fixed this a looong long time ago. Do you want to open a separate issue about symmetry?

musically-ut commented:
@amueller Sure, I can do that. 

IIRC, I wasn't sure what the option `y` was doing and we were facing a dearth of test cases which would fail in one case while work in the other, to avoid regression.

amueller commented:
We usually try to avoid Greek letter parameters and prefer natural language parameters, though not everywhere (if I'd rewrite the linear models, I'd call alpha "regularization"). I don't have a strong opinion though.

The question is whether the parameters is called ``alpha`` and the docs say ``clamping factor`` or the parameter is called ``clamping_factor`` and the docs say ``alpha in the literature``.
Actually "clamping factor" is a pretty bad description because it's the inverse of clamping. Maybe "spreading factor"?

musically-ut commented:
@amueller `spreading_factor` sounds okay. However, maintaining backwards compatibility in this case will become slightly tricky.

amueller commented:
Deprecating parameters is not really tricky, but I'm fine with keeping alpha. I was also thinking about the docs, which currently say "clamping factor" which is bad.

boechat107 commented:
I'm going to apply the suggestions regarding the deprecation of `alpha`, but I'll leave it like `alpha` for now.

boechat107 committed:
Improving "deprecated" doc string and raising DeprecationWarning

boechat107 commented:
Now that the tests are failing, I remember why I didn't use `alpha=None`.

Suggestions?

amueller commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -342,6 +343,22 @@ class LabelPropagation(BaseLabelPropagation):
+                "Deprecated in 0.19 and it's going to be removed in 0.21.",
+                DeprecationWarning
+            )
+        alpha = .0
```
don't put this here. Put it where it's used

boechat107 replied:
Ok, I'll set the `alpha` of the line below to `.0`.

But what about the `alpha` of the constructor above? You said it should be set to `None`, but the tests fail because the two constructors have different default values (`None` and `0`).

What am I missing?

amueller replied:
It shouldn't be changed in the constructor, it should be replaced in ``fit``

amueller replied:
It's a bit unfortunate but it looks like ``LabelPropagation`` needs to get it's own ``fit`` which calls
``super(LabelPropagation, self).fit``. But before, it checks whether alpha is set and raises a deprecation warning.
Then in ``BaseLabelPropagation`` you need to introduce a local variable ``alpha`` that has the right value. One way to do this is to check ``isinstance(self, LabelSpreading)`` - which is not great.
The other would be so set ``self._alpha = 0`` in ``LabelPropagation.fit``
and do ``alpha = getattr(self, "_alpha", self.alpha)`` in ``BaseLabelPropagation.fit``.

boechat107 replied:
Considering the testing restrictions, I can't see a any better solution. I chose to adopt your second suggestion.

boechat107 committed:
Using a local "alpha" in "fit" to deprecate LabelPropagation's alpha

This solution isn't great, but it sets the correct value for alpha
without violating the restrictions imposed by the tests.

boechat107 committed:
Removal of my development files

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -291,7 +295,7 @@ class LabelPropagation(BaseLabelPropagation):
 
     alpha : float
-        Clamping factor
+        DEPRECATED: Deprecated in 0.19 and it's going to be removed in 0.21.
```
can use sphinx's `.. deprecated`

boechat107 replied:
OK, I'll check it out.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -358,6 +362,18 @@ class distributions will exceed 1 (normalization may be desired).
+        # theoretical meaning (from the reference paper). Look at PR 6727.
+        if self.alpha is not None:
+            warnings.warn(
+                "Deprecated in 0.19 and it's going to be removed in 0.21.",
```
State "Parameter alpha was ... and will be ..."

boechat107 replied:
I agree.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -358,6 +362,18 @@ class distributions will exceed 1 (normalization may be desired).
+            warnings.warn(
+                "Deprecated in 0.19 and it's going to be removed in 0.21.",
+                DeprecationWarning
+            )
```
Should we be setting _alpha to the user-provided alpha until deprecation is compeleted? Otherwise our deprecation warning should state that alpha is being ignored.

boechat107 replied:
In the case of `LabelPropagation`, setting a different `alpha` value is not something supported by the algorithm's description in the cited references.

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -358,6 +362,18 @@ class distributions will exceed 1 (normalization may be desired).
+    def fit(self, X, y):
+        # alpha is deprecated for LabelPropagation because it doesn't have any
+        # theoretical meaning (from the reference paper). Look at PR 6727.
+        if self.alpha is not None:
```
you need to implement `__init__` such that `alpha=None` by default and this message only appears when a non-default value is used.

boechat107 replied:
To follow your suggestion and pass the tests, I would need to change the `__init__` of `BaseLabelPropagation`, which is probably not what we want since `LabelSpreading` can really use different `alpha` values.

This problem was discussed with @amueller before, in his code review.

jnothman replied:
I don't see how. I have no problem putting a new `__init__` in `LabelPropagation` which passes alpha=None to `BaseLabelPropagation.__init__`, while still passing the tests.

jnothman reviewed (commented):
Ideally, we would have a test that the DeprecationWarning fires when it is meant to and when not.

Please add a bug fix entry to what's new.

boechat107 committed:
Using sphinx's "deprecated" tag (jnothman's suggestion)

boechat107 committed:
Deprecation warning: stating that the alpha's value will be ignored

jnothman commented on sklearn/semi_supervised/label_propagation.py:
```diff
@@ -381,7 +402,10 @@ class LabelSpreading(BaseLabelPropagation):
 
     alpha : float
-      clamping factor
+      Clamping factor [0, 1], it specifies the relative amount of the
```
A bit more verbose for clarity: "Clamping factor, in [0, 1], specifies the extent to which a sample's label should derive from its neighbors in preference to its initial label."

boechat107 replied:
By a new `__init__` do you mean something like [this](https://github.com/scikit-learn/scikit-learn/pull/6727/commits/f6091051874f92674d609f79757f30fcbba3ab49)? 

What should be the default value for `alpha` in `BaseLabelPropagation.__init__`? What about `LabelPropagation.__init__`? The tests were failing because I ended up with `alpha=1` in `BaseLabelPropagation.__init__` and `alpha=None` in `LabelPropagation.__init__`, i.e., this constructor set a different default value for `alpha` than the former.

jnothman reviewed (commented):
I'd like to see this merged in the coming release!

jnothman commented:
I've not yet looked at test failures or otherwise, but I'd leave alpha=None
until fit, or even just remove it from base's init if it's not relevant to
all subclasses

On 29 May 2017 8:53 pm, "Andre Boechat" <notifications@github.com> wrote:

> *@boechat107* commented on this pull request.
> ------------------------------
>
> In sklearn/semi_supervised/label_propagation.py
> <https://github.com/scikit-learn/scikit-learn/pull/6727#discussion_r118913315>
> :
>
> > @@ -381,7 +402,10 @@ class LabelSpreading(BaseLabelPropagation):
>        parameter for knn kernel
>
>      alpha : float
> -      clamping factor
> +      Clamping factor [0, 1], it specifies the relative amount of the
>
> By a new __init__ do you mean something like this
> <https://github.com/scikit-learn/scikit-learn/pull/6727/commits/f6091051874f92674d609f79757f30fcbba3ab49>
> ?
>
> What should be the default value for alpha in
> BaseLabelPropagation.__init__? What about LabelPropagation.__init__? The
> tests were failing because I ended up with alpha=1 in
> BaseLabelPropagation.__init__ and alpha=None in LabelPropagation.__init__,
> i.e., this constructor set a different default value for alpha than the
> former.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6727#discussion_r118913315>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6_f4VqVwRMB0AHMXG9HOe0Yxuhq-ks5r-qOogaJpZM4IRT8O>
> .
>

jnothman commented:
The point is that you should still not have to modify anything in __init__, and leave warning until fit.

boechat107 commented:
If I don't modify `__init__`, the warning on `LabelPropagation.fit` will be echoed by default. Is this what you meant?

jnothman commented:
i was suggesting you do modify init but don't modify the parameter before
setting the attribute

On 30 May 2017 2:51 pm, "Andre Boechat" <notifications@github.com> wrote:

> If I don't modify __init__, the warning on LabelPropagation.fit will be
> echoed by default. Is this what you meant?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6727#issuecomment-304774132>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-jkNQC4VcQZUwnP5v6yjGMkog-eks5r-6AygaJpZM4IRT8O>
> .
>

jnothman committed:
Use __init__ with alpha=None

jnothman committed:
Merge branch 'master' into lpalpha

jnothman committed:
Update what's new

boechat107 committed:
Merge pull request #2 from jnothman/lpalpha

Changes to fixing #5774 (label clamping)

jnothman commented:
It seems you'll need to resolve a merge conflict.

I've now patched this to my liking. Could we get at least one other review? @amueller?

boechat107 commented:
Yes, I noticed. I'm gonna fix it today.

boechat107 committed:
Merge branch 'master' into issue-5774

jnothman commented on sklearn/semi_supervised/tests/test_label_propagation.py:
```diff
@@ -59,3 +62,15 @@ def test_predict_proba():
+    X, y = make_classification(n_samples=100)
+    y[::3] = -1
+    lp_default = label_propagation.LabelPropagation()
+    lp_default_y = assert_no_warnings(lp_default.fit, X, y).transduction_
```
This is getting a warning (although not the one we're concerned about) and the assertion fails:
```

======================================================================
FAIL: sklearn.semi_supervised.tests.test_label_propagation.test_alpha_deprecation
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/semi_supervised/tests/test_label_propagation.py", line 71, in test_alpha_deprecation
    lp_default_y = assert_no_warnings(lp_default.fit, X, y).transduction_
  File "/home/travis/build/scikit-learn/scikit-learn/sklearn/utils/testing.py", line 230, in assert_no_warnings
    ', '.join(str(warning) for warning in w)))
AssertionError: Got warnings when calling fit: [{message : RuntimeWarning('underflow encountered in exp',), category : 'RuntimeWarning', filename : '/home/travis/build/scikit-learn/scikit-learn/sklearn/metrics/pairwise.py', lineno : 837, line : None}]
```

I've realised that I didn't fix the `random_state` in `make_classification`. We should do that. But I also haven't investigated why we should be getting exp underflow. Do you know?

boechat107 replied:
It is not something that comes to mind right now. I'll have look at it.

jnothman replied:
I don't seem to get that warning on my system, but all Travis runs get it.

Could I suggest you try setting the `kernel='knn'` in the test here, and we'll see if that works?

jnothman commented:
I've made a couple of finishing touches at #9192.

ogrisel commented:
Closing in favor of #9239.
